src/charm.py

```diff
@@ -1,117 +1,296 @@
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        self._tls.certificate_relation_available(event)
+        self._reconcile()
+
+    def _on_certificate_expiring(
+        self,
+        event: Union[CertificateExpiringEvent, CertificateInvalidatedEvent],
+    ) -> None:
+        """Handle the TLS Certificate expiring event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        if not self._are_relations_ready(event):
+            return
+        self._tls.certificate_expiring(event, self.certificates)
+
+    # This method is too complex but hard to simplify.
+    def _certificate_revoked(self, revoke_list: List[str]) -> None:  # noqa: C901
+        """Handle TLS Certificate revocation.
+
+        Args:
+            revoke_list: TLS Certificates list to revoke
+        """
+        if not self._are_relations_ready(None):
+            return
+        tls_certificates_relation = self._tls.get_tls_relation()
+        for hostname in revoke_list:
+            old_csr = self._tls.get_relation_data_field(
+                f"csr-{hostname}",
+                tls_certificates_relation,  # type: ignore[arg-type]
+            )
+            if not old_csr:
+                continue
+            if JujuVersion.from_environ().has_secrets:
+                try:
+                    secret = self.model.get_secret(label=f"private-key-{hostname}")
+                    secret.remove_all_revisions()
+                except SecretNotFoundError:
+                    LOGGER.warning("Juju secret for %s already does not exist", hostname)
+                    continue
+            try:
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    tls_certificates_relation,  # type: ignore[arg-type]
+                )
+                peer_relation = self.model.get_relation("nginx-peers")  # type: ignore[arg-type]
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    peer_relation,  # type: ignore[arg-type]
+                )
+            except KeyError:
+                LOGGER.warning("Relation data for %s already does not exist", hostname)
+            self.certificates.request_certificate_revocation(
+                certificate_signing_request=old_csr.encode()
+            )
+
+    def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
+        """Handle the TLS Certificate invalidation event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
```

Comment:
Analogous case here

Reply:
I'm marking this as resolved as per my comment above.